keps/sig-storage/20190122-csi-inline-volumes.md

```diff
@@ -225,19 +228,39 @@ Since ephemeral volume requests will participate in only the mount/unmount volum
 * Volume operation that use CSIVolumeSource can only work with proper feature gate enabled
 
 ### Ephemeral inline volumes unit tests
-* Ensure required fields are provided
-* Mount/Unmount should be triggered with CSIVolumeSource
-* Expected generated volumeHandle is created properly
+* Ensure required fields are provided: csi.storage.k8s.io/ephemeral (https://github.com/pohly/kubernetes/blob/4bc5d065c919fc239e2c8b40e6a96e409ca011fd/pkg/volume/csi/csi_mounter_test.go#L140-L146)
+* Mount/Unmount should be triggered with CSIVolumeSource: https://github.com/kubernetes/kubernetes/blob/10005d2e1e1425904f8c7bf5615e730fb0fea7c9/pkg/volume/csi/csi_mounter_test.go#L386
+* Expected generated volumeHandle is created properly: https://github.com/kubernetes/kubernetes/blob/10005d2e1e1425904f8c7bf5615e730fb0fea7c9/pkg/volume/csi/csi_plugin_test.go#L177
+* Ensure that CSIDriver.Spec.Mode field is validated properly: https://github.com/kubernetes/kubernetes/pull/80568
 * Ensure volumeHandle conforms to resource naming format
 * CSIVolumeSource info persists in CSI json file during mount/unmount
 * Ensure Kubelet skips attach/detach when `CSIDriver.Mode = ephemeral`
-* Ensure Kubelet skips inline logic when `CSIDriver.Mode = persistent` or `CSIDriver.Mode is empty` 
+* Ensure Kubelet skips inline logic when `CSIDriver.Mode = persistent` or `CSIDriver.Mode is empty`: covered by existing tests
 
 ### E2E tests
-* Pod spec with an ephemeral inline volume request can be mounted/unmounted
-* Two pods accessing the same ephemeral inline volumes
-* Single pod referencing two distinct inline volume request from the same driver 
+* Pod spec with an ephemeral inline volume request can be mounted/unmounted: https://github.com/pohly/kubernetes/blob/4bc5d065c919fc239e2c8b40e6a96e409ca011fd/test/e2e/storage/csi_mock_volume.go#L356-L371, https://github.com/pohly/kubernetes/blob/4bc5d065c919fc239e2c8b40e6a96e409ca011fd/test/e2e/storage/testsuites/ephemeral.go#L110-L115
+* Two pods accessing the same ephemeral inline volume
```

Comment:
How about:
Two pods configured the same (.spec.volumes.name, driver, args) get different volumes.

Reply:
What is the definition of "different volumes"?

I changed the test case description to "two pods accessing an ephemeral inline volume which has the same attributes in both pods" and the corresponding test creates a second pod, but I don't have any sanity testing for "different volumes".

Reply:
@kfox1111 we can still work on the test cases later. This shouldn't block merging this KEP.

Reply:
For example, say you made pod A that has a hostpath ephemeral volume named 'foo' and a command [/bin/sh, -c, 'hostname > /ephemeralmountpoint/hello.txt']

You could wait until you saw that pod go up and the hello.txt get created.

You could then create a second pod, on the same host with a volume named 'foo' configured the same way with: 
command [/bin/sh, -c, 'while true; do sleep 1000; done']
And see if you can read /ephemeralmountpoint/hello.txt out of the second container?

Reply:
So "different volumes" basically means "changes not propagated". Yes, that can be tested.

In the meantime, are you okay with merging this PR?

Reply:
Now that I started to think about it... who guarantees that two pods with the same (as in volume attributes) volume end up with separate volumes? The entire semantic of those volumes is left to the CSI driver. For example, a CSI driver might take no attributes and connect all pods to the same NFS share. That wouldn't be a very useful driver, but it wouldn't be invalid.

I think the test that I have written so far is useful (covers a potential corner case), but we can't make it more strict and keep it generic at the same time.

Reply:
the ephemeral volumeHandle generation mechanism ensures that two different pods with the same exact config on the same host, or the same pod with the same config but different volume names each are different volumeHandles.

Reply:
For example, very psudocody:
```
kind: Deployment
spec:
  replicas: 2
  spec:
    nodeSelector: samenode
    volumes:
    - name: foo
      csi:
        driver: hostpath.k8s.io
```
Creates two different volumes, one per pod. Each pod's volumeHandle is different.

This one also creates two different backing volumes. Two different volumeHandles:
```
kind: Pod
spec:
  volumes:
  - name: foo
    csi:
      driver: hostpath.k8s.io
  - name: bar
    csi:
      driver: hostpath.k8s.io
```

In each case, with the ephemeral hostpath driver backing it, it would be the equivalent behavior of using emptyDir: {} instead of csi. No sharing between volumes.

Your technically right, the volume physically being different could be an implemntation detail of the driver. So it could share a read only filesystem and do some careful reference counting. But the above two tests are really the ones I'm interested in making sure work. If they work, other drivers will work properly. Its basically, does the ephemeral volumeHandle generation work properly all the way to the driver. Then the driver can do what it wants with that.

Reply:
So you want to test volume handle generation, not the behavior of the resulting volumes.

Reply:
Yeah.